Addition of charts and graphs

scottkellum opened the pull request:
## Description

This adds the ability to easily generate simple charts with tables. Simply put your content into a well formatted table and add the class `.usa-chart`. The goal was to make this as simple and easy to use as possible while also remaining accessible to those without the use of JavaScript.

## What problem does this solve?

Agencies already use graphs to display information to audiences and this allows them to embed them directly into their sites more easily and with better accessibility than images.

## Additional information

### Bar chart example

<img width="597" alt="screen shot 2018-01-29 at 2 42 52 pm" src="https://user-images.githubusercontent.com/377189/35530710-dd4ca328-0502-11e8-8f9f-912bd959287c.png">

```html
<div class="usa-chart" data-chart-type="bar"> <!-- char-type: bar, line -->
  <table class="usa-chart-data" data-min="0" data-max="100" data-increment="20" data-grid="true"> <!-- all values optional -->

    <!--
    label datasets with header row
    unnessisary unless multiple values are assigned
    -->
    <tr>
      <td></td>
      <th class="usa-chart-dataset">set one</th>
      <th class="usa-chart-dataset">set two</th>
    </tr>

    <tr class="usa-chart-item">
      <td class="usa-chart-label">Thing one</td>
      <td class="usa-chart-value">8%</td>
      <td class="usa-chart-value">28%</td>
    </tr>
    <tr class="usa-chart-item">
      <td class="usa-chart-label">Thing two</td>
      <td class="usa-chart-value">12%</td>
      <td class="usa-chart-value">24%</td>
    </tr>
    <tr class="usa-chart-item">
      <td class="usa-chart-label">Thing three</td>
      <td class="usa-chart-value">26%</td>
      <td class="usa-chart-value">40%</td>
    </tr>
    <tr class="usa-chart-item">
      <td class="usa-chart-label">Thing four</td>
      <td class="usa-chart-value">60%</td>
      <td class="usa-chart-value">42%</td>
    </tr>
  </table>
</div>
```

### Line graph example

<img width="595" alt="screen shot 2018-01-29 at 2 43 09 pm" src="https://user-images.githubusercontent.com/377189/35530742-f9de333a-0502-11e8-968c-2687d76a6ed2.png">

```html
<div class="usa-chart" data-chart-type="bar"> <!-- char-type: bar, line -->
  <table class="usa-chart-data" data-min="0" data-max="100" data-increment="20" data-grid="true"> <!-- all values optional -->

    <!--
    label datasets with header row
    unnessisary unless multiple values are assigned
    -->
    <tr>
      <td></td>
      <th class="usa-chart-dataset">set one</th>
      <th class="usa-chart-dataset">set two</th>
    </tr>

    <tr class="usa-chart-item">
      <td class="usa-chart-label">Thing one</td>
      <td class="usa-chart-value">8%</td>
      <td class="usa-chart-value">28%</td>
    </tr>
    <tr class="usa-chart-item">
      <td class="usa-chart-label">Thing two</td>
      <td class="usa-chart-value">12%</td>
      <td class="usa-chart-value">24%</td>
    </tr>
    <tr class="usa-chart-item">
      <td class="usa-chart-label">Thing three</td>
      <td class="usa-chart-value">26%</td>
      <td class="usa-chart-value">40%</td>
    </tr>
    <tr class="usa-chart-item">
      <td class="usa-chart-label">Thing four</td>
      <td class="usa-chart-value">60%</td>
      <td class="usa-chart-value">42%</td>
    </tr>
  </table>
</div>
```

maya commented:
This is looking good! We've heard charts and graphs are something many agencies would need (see #18). We're still figuring out what our component submission workflow will look like and will let you know when we have more info. Appreciate the simplicity and focus on progressive enhancement here.

scottkellum commented:
@maya I’m feeling comfortable for this to be reviewed now when y’all are ready.

- I removed the pie chart because [they aren’t very good at displaying information](http://www.businessinsider.com/pie-charts-are-the-worst-2013-6)
- Consolidated everything into a single component to reduce duplication
- Added support for more than one dataset on a graph

maya commented:
Thanks so much @scottkellum! We'll probably be able to start reviewing this (in _some_ way) in the next sprint (or the one following that), as this sprint we're focused on building out utilities, and we've tentatively planned the next sprint to be a technical debt sprint. The process may take some time as you're our first major component submission, so we appreciate the patience.

Breaking up the review for the team, which could happen async:

- [ ] UX -- @donjo 
- [ ] Visual design -- @thisisdano 
- [ ] Front-end
   - [ ] HTML, CSS, a11y -- @maya or @donjo 
  - [ ] JS -- [TBD engineer]

---
For this to go on the main website, we'll need documentation.

Open question:
Is this included in the main package, behind a `experimental` flag that's not included in the package unless you opt-in, or through CodePen?

scottkellum commented:
Thanks @maya ! Let me know if you need anything from me in the process and I’ll sit tight as you move through your sprints.

maya commented:
Hey @scottkellum, could you fix the linting errors in CircleCI? You can also run `npm test` in your code.

scottkellum commented:
@maya Yup! Sorry about that should pass now

maya commented on src/components/11-infographics/chart.njk:
```diff
@@ -0,0 +1,35 @@
+<div class="usa-chart" data-chart-type="bar"> <!-- char-type: bar, line -->
+  <table class="usa-chart-data" data-min="0" data-max="100" data-increment="20" data-grid="true"> <!-- all values optional -->
+
```
Can you add a `<caption>` as the title giving a brief description of the table? See: https://webaim.org/techniques/tables/data#caption

maya commented on src/components/11-infographics/chart.njk:
```diff
@@ -0,0 +1,35 @@
+    -->
+    <tr>
+      <td></td>
+      <th class="usa-chart-dataset">set one</th>
```
In various places: can you add the appropriate scope attributes to identify your headers? See: https://webaim.org/techniques/tables/data#headers

scottkellum commented:
I’m working on this feedback, I need to figure out a way to surface the `caption` in the rendered graph.

In terms of accessibility please review the rendered SVG as well. I tested it with VoiceOver and it reads through the data appropriately but you might have insights on things I’m missing.

donjo commented:
Hey @scottkellum, got a chance to look at this a bit more thoroughly today and offer some general UX/design feedback. I'll mention ahead of time that a lot of this is based on things I learned in working on a mini discovery project with census (github.com/18F/uswds-data) where we were trying to establish the baseline of some data design guidelines. That said here are some suggestions and I'm curious to hear what you think. 

**Add titles and source placeholder text**
- One thing the Census team kept driving home was the importance of linking to data sources. While it won't be necessary 100% of the time, it could be useful to include a placeholder here to nudge users into including a link to the raw data if they have it. As for titles, looks like @maya already has that covered. To go with that I'd suggest moving the chart keys to below the chart

**Separate `njk` for line chart**
- Would be helpful for folks looking at the component library to see there are multiple options without having to read the comments. 

**Adjust font size on label**
- May just be a personal preference but the labels seem a bit large and overpowering. What do you think about going from 4.25px to 4px in 

```
.usa-chart-gridLabel {
+  text-anchor: middle;
+  font-size: calc(4.25px - .25vw);
}
``` 

I think that's all I got for now. Great work!

maya commented:
@scottkellum I thought when I previously tested the rendered SVG it read it like the table, in that I was able to progress through each value one by one, but now it seems to read it all smushed together. Was that intentional?

<img width="1280" alt="screen shot 2018-02-20 at 2 27 36 pm" src="https://user-images.githubusercontent.com/5249443/36452727-71c8bd04-164a-11e8-87c7-1cba248854f4.png">

---
For the title, it seems like it should render a `<title>` in the SVG and a `<caption>` in the table.

maya commented:
For the font size, could we just use the `$base-font-size`?

maya commented:
Oh, nevermind about the font-size. Looks like it's responsive to the size of the chart 👍.

scottkellum commented:
@maya @donjo Thanks for your feedback! I made revisions and feel like the accessability is much improved. For a chart title I just dropped an `h2`.

maya commented:
Thanks @scottkellum! 

I've talked with a few people about this and heard the best practice is to make the SVG hidden for screen readers and include the table. The table would be visually hidden but accessible to screen readers and you could have the option to show it visibly too along with the chart.

Karl Groves mentioned including tables allows you to identify table headers and explore the table using typical table reading commands.

Some background from my colleague @toolness: 
>as far as i can tell, [the best practice] is to provide a table as additional content, so that the bar chart is basically just an image (though its general shape could be provided as alt text) but the user can deep-dive into the data of the table
>
>heydon pickering gave a talk where he actually argues that showing the table _only_ to certain users is actually non-inclusive, b/c even some sighted users (like himself) prefer tables to bar charts
>
>i find that SVGs are very unpredictable with screen readers--some readers essentially treat them opaquely like image elements, while others actually let you navigate through individual elements of the SVG

scottkellum commented:
Thanks @maya! I was using [info from CSS-Tricks](https://css-tricks.com/accessible-svgs/) when making these accessible but this makes a lot more sense. I’ll try to get to this today.

scottkellum commented:
@maya I’m playing with this and it doesn’t feel like it would be a great experience if the user has low vision. The visual indicators on screen are no longer aligned with the text. Navigation does feel much better for users who can’t use the screen.

Updates pushed though, let me know how these feel to you.

maya commented:
I see what you're saying and am checking if that's an issue. A little more feedback I've gotten:

Someone had mentioned using a toggle to show the table instead of a chart. For someone using a magnifier, the chart may be harder to read than the table. And he half-expected the link to data sources to show the table. 

Whether there's a toggle to replace the chart with the table or just show both may depend on the audience and am not sure which approach is the best way.

scottkellum commented:
This is really good feedback @maya thank you! I think I’m going to take a step back and diverge a bit thinking about the various ways to display this data and giving both authors and users more flexibility in how that data is presented.

maya commented:
I've shared this with our Section 508 listserv and am getting some helpful responses (will keep them coming as I get them):

---
Maya,

I know it’s not quite the question, but I’m sighted and often prefer a table to a graph. Also, if I really want to get into the data, if there’s a table, I can copy it into Excel and sort, filter, etc.

So, that’s the general usability argument for keeping the table for everyone. As for accessibility, people with low vision or cognitive disabilities might be negatively affected by not having access to the table.

People with low vision might also want to copy and paste into Excel to have more control over what they see. Some people with low vision have a very small field of vision, so looking at the middle of a bar chart could leave them without the ability to see the X or Y axis. A table might be easier to follow, even if the header row and first column are out of the field of vision. I don’t know, but it’s probably worth investigating. Also, some people with low vision have to zoom to 200% or more—SVG is a great start, but that can still put the X and Y axes out of view.

I also imagine there must be cognitive disabilities for which a table is preferable. I know for myself that the likelihood that I’d prefer a table goes up with the noise or other distractions around me, so perhaps people with attention-related disorders?

They’re just thoughts, really, but my immediate gut reaction to your idea was “No! Don’t take my table away!” Charts are great for most of us for a quick look, but as soon as you start really thinking about the data, tables are just way easier to use.

Bethany Letalien
Section 508 Coordinator
U.S. Office of Personnel Management
Office of the Chief Information Officer

---

Maya,

I prefer your last two display options: (1) show the table below the graph or (2) add a toggle to show the table instead of the graph.

The reasons:

- You do not want actionable hidden content (e.g. the first “data sources” link) to be accessible via the keyboard, because keyboard navigation is not always paired with using a screen reader to hear the hidden content.  Users may access content using only the keyboard for a variety of reasons, and a visual indication of where the keyboard focus is currently located is important.
- People absorb information in different ways. If the data is already prepared in both graph and table format, why not allow all users to access both?
- Maintenance… Hidden content often gets overlooked when updates are made.

Side note:  I’m using IE11. When I open your example link, the bar chart is empty and I do not see a table. If I view in Compatibility Mode, then the chart disappears and I only see the table.  It could be my locked down environment, or there could be some browser troubleshooting needed on your end.  Good luck!

Best wishes,
-Dawn

Dawn Wyse Hurto
Enterprise Business Solutions  |  Office of the Chief Information Officer  |  U.S. Department of the Treasury
DHS Section 508 Trusted Tester - Certification Number 300297

scottkellum commented:
Hey @maya, Unfortunately I don’t have bandwidth to work on this right now or in the near future.

maya commented:
Ah OK, thanks so much for the work so far! Closing this PR for now until we have the bandwidth to pick it up and continue building on this great work.
